src/tracing/core/shared_memory_arbiter_impl.cc

```diff
@@ -908,6 +908,23 @@ void SharedMemoryArbiterImpl::NotifyFlushComplete(FlushRequestID req_id) {
   }
 }
 
+void SharedMemoryArbiterImpl::ScrapeEmulatedSharedMemoryBuffer(
+    const std::map<WriterID, BufferID>& buffer_for_writers) {
+  PERFETTO_CHECK(use_shmem_emulation_);
+
+  ForEachScrapableChunk(
+      &shmem_abi_, [&](SharedMemoryABI::Chunk* chunk, auto, auto) {
+        const auto writer = buffer_for_writers.find(chunk->writer_id());
+        if (writer == buffer_for_writers.end())
+          return;
+
+        BufferID target_buffer_id = writer->second;
+        PatchList ignored;
+
+        ReturnCompletedChunk(std::move(*chunk), target_buffer_id, &ignored);
```

Comment:
Hmm why are we returning the Chunk as completed? This is what worries me here:

1. Somewhere in the same VM you have a TraceWriter which owns this chunk of SMB (well the TW thinks it's shared, it doesn't know it's not)
2. Now when you get a flush, you return the completed chunk.
3. This is fine from a service viewpoint, but from the client viewpoint this will trigger bugs, i believe, because the the TraceWriter it's still convinced it owns the chunk, and it will keep writing onto that (And eventually it will call ReturnCompletedChunk itself).

I think here you should only send over IPC a copy of the chunk to the service, without marking it as complete in the local SMB. the TW needs to be able to keep writing into that, unless i'm missing something